llvm/lib/Target/AMDGPU/SIPreEmitPeephole.cpp

```diff
@@ -304,11 +299,62 @@ bool SIPreEmitPeephole::getBlockDestinations(
   return true;
 }
 
+namespace {
+class BranchWeightCostModel {
+  const SIInstrInfo &TII;
+  const TargetSchedModel &SchedModel;
+  BranchProbability BranchProb;
+  static constexpr uint64_t BranchNotTakenCost = 1;
+  uint64_t BranchTakenCost;
+  uint64_t ThenCyclesCost = 0;
+
+public:
+  BranchWeightCostModel(const SIInstrInfo &TII, const MachineInstr &Branch,
+                        const MachineBasicBlock &Succ)
+      : TII(TII), SchedModel(TII.getSchedModel()) {
+    assert(SchedModel.hasInstrSchedModelOrItineraries());
+
+    const MachineBasicBlock &Head = *Branch.getParent();
+    const auto *FromIt = find(Head.successors(), &Succ);
+    assert(FromIt != Head.succ_end());
+
+    BranchProb = Head.getSuccProbability(FromIt);
+    assert(!BranchProb.isUnknown());
```

Comment:
Is it actually a problem if it's unknown?

Reply:
The problem with the unknown case is that we get as a probability something bigger than `1`: "UINT32_MAX / 2**31". Then, the subtraction `Denominator - Numerator` performed later would underflow.

Reply:
But does it matter that it underflowed? The logic will still come out to return false? This check is not required for correctness

Reply:
I agree that it doesn't matter if it underflows; we never hit this case anyways.

But if we hit it, I think it would handle the transformation as profitable.
It should be similar to handling the branch with a highly likely probability (bigger than 1) of taking the edge towards the Then block.